builtin/joined_dataset.cc

```diff
@@ -59,6 +59,30 @@ struct JoinedDataset::Itl
         //ML::compact_vector<RowHash, 2> rowHashes;   ///< Row hash from input datasets
     };
 
+     struct JoinedRowStream : public RowStream {
+
+        JoinedRowStream(JoinedDataset::Itl* source) : source(source)
+        {
+            
+        }
+
+        virtual std::shared_ptr<RowStream> clone() const{
+            auto ptr = std::make_shared<JoinedRowStream>(source);
+            return ptr;
+        }
+
+        virtual void initAt(size_t start){
+            iter = source->rows.begin() + start;
+        }
+
+        virtual RowName next() {
```

Comment:
Instead of next(), I would create next_n() which returns more than one in an array you pass in.  That will amortize the cost of the virtual method call, which could be quite significant here.